### PR TITLE
Topic recognition

### PR DESCRIPTION
**Pull request by huashu20041017**
Here is an implementation of the key components in Python for the task of implementing, training, and testing a model using the files: modules.py, dataset.py, train.py, and predict.py. This will give a complete structure, including dataset loading, model definition, training, and prediction.

### CODE REVIEW

**Comment by huashu20041017** [comment]
modules.py: This file contains the components of the model (for example, a simple CNN).
dataset.py: This file contains the data loading and preprocessing pipeline.
train.py: This file contains the training, validation, and testing loop.
predict.py: This file shows how to use the trained model to make predictions on new data.

**Comment by huashu20041017** [comment]
Change the file location to the correct folder

**Comment by shaivikaaaa** [comment]
@gayanku

**Comment by wangzhaomxy** [comment]
This is a redundant pull request with the same content as PR #12.

I won't be marking again here.

**Comment by wangzhaomxy** [comment]
I realize you changed the merge branch in this PR, which is the right one. I will copy the comment from #12. This PR will be well merged into Shakes' branch. 

Please provide your student number. Otherwise, we cannot give you any marks.

**Comment by wangzhaomxy** [comment]
**<This is an initial inspection, no action is required at this point.>**

**File Organizing:** Wrong file path. All student-constructed documents (involving README.md) should be packed in the ‘PatternAnalysis-2024/recognition/<student_constructed_folder>/’ folder. Please reorganize your files, otherwise it won't be merged into Shakes' branch.

**Problem Solving:**

- The algorithm solves the problem appropriately. 
- Accuracy in the testing dataset: 0.75. However, no evidence supports this result in the report. In addition, the student mentioned the Training and Validation results using the Dice coefficient, which is not appropriate. Furthermore, 0.78 and 0.76 Training/val dice scores are rare in the UNet training phase.

**Model and functions:**

- It correctly uses PyTorch to construct the 2D-UNet models and functions.
- No data augmentation.

**Code design:** Good.

**Code comment and docstring:**

- Good code comments 
- No function docstrings

**Difficulty:** Easy.

**Additional Comments:**

- Insufficient and meanless commits, such as "update xxxx". Most of the commits are copy-and-paste code.
- Poor report format: ChatGPT style. The sentence in the third line of the Dataset part, "Dataset Source:[Provide link if applicable]", is suspected ChatGPT generated content. 
- No figure in the report. The experiment training, validation, and testing process are expected to be shown by the figures in the report.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. No evidence supports acc result in the report.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting missing. No doc strings.</td><td>-1</td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Easy. UNet2D (Easy Difficulty)</td><td>-10</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. </td><td>-1</td></tr>
	<tr><td></td><td>Some/Adequate Progressive commits. Commits lack purpose.</td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. No images/Charts etc as evidance of model training and predictions.</td><td>-4</td></tr>
	<tr><td></td><td>Model/technical explanation :Acceptable. </td><td>-2</td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Pull Request has problems: Submitted on time. Wrong branch.</td><td>-2</td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is adequate. </td><td>-1</td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-24</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by hanemma7moud** [comment]
Hua Shu
